Limit number of celery task executions per second per user

claudiofr opened the pull request:
Closes #14411

Introduce a celery_user_rate_limit config parameter which specifies how many tasks per second a user can execute. Create a custom celery Task class called GalaxyTask with a before_start hook that implements the rate limiting logic. Add a new table, celery_user_rate_limit that tracks the last scheduled execution time
by user id. Add new integration test, test_celery_user_rate_limit.py.

A task function must include the keyword parameter task_user_id for user rate limiting to work.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

jdavcs commented on lib/galaxy/app.py:
```diff
@@ -576,6 +582,25 @@ def __init__(self, configure_logging=True, use_converters=True, use_display_appl
+        """
+        Register subtype class instance to support implementation of a user rate limit for execution of celery tasks.
+        The default supertype class does not enforce a user rate limit. This is the case if the celery_user_rate_limit
+        config param is default value of 0.0
```
Maybe not specify `0.0` as the default value in this comment? It's already specified in the config schema, and if we choose to modify it we'd have to remember to modify it here as well (we will forget, of course).

claudiofr replied:
Do you want me to undo the manual changes to *.sample and *.rst at this point and redo them using make config-rebuild? Also as far as moving migration to a separate commit do you mean to simply edit the migration script and then push it by itself to the same branch under the same PR?

jdavcs replied:
Yes, please! Both can be separate commits in the same PR.

jdavcs commented on lib/galaxy/app.py:
```diff
@@ -576,6 +582,25 @@ def __init__(self, configure_logging=True, use_converters=True, use_display_appl
+        """
+        if self.config.celery_user_rate_limit:
+            force_standard_before_start = self.config.get("celery_user_rate_limit_standard_before_start", False)
+            if not force_standard_before_start and self.config.database_connection.startswith("postgres"):
```
```suggestion           
            if not self.config.celery_user_rate_limit_standard_before_start and self.config.database_connection.startswith("postgres"):
```
We don't need to use `get` to access a config value: any value `foo` defined in the schema can be accessed here as `self.config.foo`; if the `foo` option is not set, it'll contain the default value as per its schema definition.
If you'd rather keep `force_standard_before_start` as the variable name, that's fine.

jdavcs commented on lib/galaxy/celery/base_task.py:
```diff
@@ -0,0 +1,69 @@
+                usr, sa_session, self.task_exec_countdown_secs, now
+            )
+        finally:
+            if sa_session:
```
I think this will always evaluate to `True` because the line `sa_session = self.ga_scoped_session` will always succeed (even if the next line raises an error), and, unless I'm mistaken, `self.ga_scoped_session` is never `None`.

jdavcs commented on lib/galaxy/celery/base_task.py:
```diff
@@ -0,0 +1,69 @@
+            )
+        finally:
+            if sa_session:
+                sa_session.remove()
```
No need to remove: the session is automatically closed at the end of the request (this change is very recent). (and we shouldn't close it prematurely)

claudiofr replied:
Just curious where is the logic that automatically closes the session? When you say it is closed at the end of the request do you mean the web api request?

jdavcs replied:
Yep. The session is closed here: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/model/base.py#L113
It's called at the end of each request:
- https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/api/__init__.py#L95 (for fastapi)
- https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/web/framework/base.py#L175 (for wsgi endpoints)

jdavcs commented on lib/galaxy/model/__init__.py:
```diff
@@ -10233,6 +10237,111 @@ class CleanupEventImplicitlyConvertedDatasetAssociationAssociation(Base):
+        user_id=bindparam("userid"), last_scheduled_time=bindparam("sched_time")
+    )
+
+    def calculate_task_start_time(
```
This applies to both of the `calculate_task_start_time` methods.

Since the executed statement writes to the database, we need to explicitly commit, otherwise a rollback will be issued on session close (the session is no longer autocommited after #15421).

You can use this code: (the use of the context manager is a precaution during the transition to no autocommit and SQLAlchemy 2.0)

```py
from galaxy.model.base import transaction
...
with transaction(sa_session):
    sa_session.commit()
```

jdavcs commented on lib/galaxy/model/migrations/alembic/versions_gxy/987ce9839ecb_create_celery_user_rate_limit_table.py:
```diff
@@ -0,0 +1,27 @@
+
+
+def upgrade():
+    op.create_table(
```
Please use the `create_table` helper from `galaxy.model.migrations.util`

jdavcs commented on lib/galaxy/model/migrations/alembic/versions_gxy/987ce9839ecb_create_celery_user_rate_limit_table.py:
```diff
@@ -0,0 +1,27 @@
+
+
+def downgrade():
+    op.drop_table("celery_user_rate_limit")
```
same as above, `drop_table`.

jdavcs commented on lib/galaxy/model/__init__.py:
```diff
@@ -10233,6 +10237,111 @@ class CleanupEventImplicitlyConvertedDatasetAssociationAssociation(Base):
+        )
+
+
+class CeleryUserRateLimitStandard(CeleryUserRateLimit):
```
This object hierarchy makes sense, of course. However, I wonder if it might be better suited for a manager, not a model? 

Here's why. According to SQLAlchemy documentation, when we have single table inheritance, a discriminator column is required ([ref](https://docs.sqlalchemy.org/en/14/orm/inheritance.html#single-table-inheritance)). Yes, here there is no distinction between the classes in terms of state (or mapped attributes) - it's just different implementation of the same method (which, again, make perfect sense). But I wonder whether the fact that these classes become mapped is not going to cause issues without specifying `polymorphic_identity` (I'm not sure). At the same time, adding that discriminator here seems silly and unnecessary. So, maybe move this logic into a manager (or managers), and keep the model focused on defining the mapped attributes? Again, I'm not at all sure about this, but it does feel like a simpler and safer alternative.

mvdbeek commented on lib/galaxy/model/__init__.py:
```diff
@@ -10233,6 +10237,111 @@ class CleanupEventImplicitlyConvertedDatasetAssociationAssociation(Base):
+            try:
+                sched_time = now
+                sa_session.execute(self._insert_stmt, {"userid": user_id, "sched_time": sched_time})
+            except Exception:
```
This should be a more narrow Exception class

claudiofr committed:
Move behavior from celery_user_rate_limit model class to base_task module.

Moved behavioral logic from CeleryUserRateLimit model class to base_task
module because model classes should be plain old classes. Ran
make config-schema to update galaxy_options.rst and galaxy.yml.sample
based on changes made to config_schema.yml.

claudiofr committed:
Limit number of celery tasks that a user can run per second.

Introduce a celery_user_rate_limit config parameter which
specifies how many tasks per second a user can execute. Create a
custom celery Task class called GalaxyTask with a before_start hook
that implements the rate limiting logic. Add a new table,
celery_user_rate_limit that tracks the last scheduled execution time
by user id. Add new integration test, test_celery_user_rate_limit.py.

claudiofr committed:
Add celery_user_rate_limit_table to rate limit task executions per user

Added migration script to create the new table which tracks the last
scheduled execution time for a user which is used to schedule the next
execution time in order to limit the task execution rate.

jdavcs commented on lib/galaxy/app.py:
```diff
@@ -581,8 +586,30 @@ def __init__(self, configure_logging=True, use_converters=True, use_display_appl
+        config param is the default value.
+        """
+        if self.config.celery_user_rate_limit:
+            task_before_start: GalaxyTaskBeforeStart
```
I think we need to register `GalaxyTaskBeforeStart` regardless of whether `celery_user_rate_limit` is set or not: we refer to it in `before_start` in `GalaxyTask`, which is specified as the base class for tasks in the `galaxy_task` decorator. Unless I'm not reading this correctly?

claudiofr replied:
With lagom, the di container you are using, there is no need to register a type if there are no constructor parameters. It does this implicitly. So the code works correctly as is. I have integration tests for this. However, it is inefficient because it will create a new instance of GalaxyTaskBeforeStart everytime it is requested in the GalaxyTask.before_start method. So in order to make the code more efficient it is better to explicitly register GalaxyTaskBeforeStart as a singleton.

jdavcs commented on lib/galaxy/config/schemas/config_schema.yml:
```diff
@@ -3701,6 +3701,23 @@ mapping:
+        required: false
+        desc: |
+          Applies if celery_user_rate_limit is non-zero. Used for testing against
+          a postgres db. Forces use of standard sql code rather than
```
Is this config option used for tests only? If so, can we keep it in the testing code and not in the config schema? (we try to limit our user-facing configuration to options relevant for Galaxy admins). I don't think an admin would need to select what SQL dialect is being used.

claudiofr replied:
Originally I did not put it in the config schema for the reason you mention. That is why I referenced in the code using the self.config.get("celery_user_rate_limit_standard_before_start", False) syntax which you commented on earlier. However, it occurred to me that there is a chance, albeit small, that a developer may in the future introduce a config parameter with the same name in the config schema. So I decided to put it in the config schema. Without a central registry of all parameters including those used only for testing there is always a chance of a naming collision. However, I'm ok with removing it from the config schema and using the self.config.get syntax to get its value.

jdavcs replied:
Ah, I see! I didn't realize at the time that it was only there for the purpose of setting up the right configuration for a test. Given that this option is only relevant in the context of a test, it (ideally) shouldn't be part of the main code base. (I say "ideally" because we still have a couple of settings like that left over from ancient times, but we try to remove those as time permits). However, you don't really need it: Galaxy can determine what path to take based on the value of `database_connection`: with a non-postgres connection a test will automatically take the standard sql path.

claudiofr replied:
I thought it would be useful to test a the standard sql path even with postgres. That is why I introduced this testing parameter. Given that it would have to be in the main code base.

jdavcs replied:
It'll never take that path: any such alternative sql implementations in galaxy are strictly postgres OR anything else (with sqlite being the default and mysql no longer formally supported, but know to work). So as long as the conn string is identified as postgres, the postgres-specific implementation should be selected.

jdavcs commented on test/integration/test_celery_user_rate_limit.py:
```diff
@@ -0,0 +1,169 @@
+
+
+@lru_cache()
+def setup_users(dburl: str, num_users: int = 2):
```
If the sole purpose of this function is to ensure we have enough users in the test db (currently, only +1 with `id=2`), we already have a `_setup_user` method that sets up a proper user. And if you only need just a row in the user table, you could you a session to simplify this to:
```py
for _ in range(num_users):
    session.add(User(whatever-args-you-need))
session.commit()
```
You don't need to explicitly set the user IDs, right?

claudiofr replied:
Having explicitly set user id's of 1 and 2 makes the testing code easier to write. However, if it is important to do so I could change the test code to lookup the user id's. In that case I would just need to ensure that there are at least 2 users defined in the user table.

jdavcs replied:
I don't think we can assume specific ID values here: the database may be shared with other tests (a trade-off we accept for the sake of faster tests). We only know for sure that there is a user.id=1; any other User records may, potentially, interfere with the correctness of these tests. Ideally, the test would create the users it needs.

claudiofr replied:
The test does not assume the presence of any user other than user id 1. It creates the other users it needs with id's 2 and up. So I don't see any problem here.

jdavcs replied:
we cannot assume that user.id=2 (or any other value) is not already taken (by another test).

claudiofr replied:
There is no problem if user.id=2 or any other user id is taken up by another test. I just need there to be a record in that table. If it already exists I don't create it. The column values don't matter and I don't manipulate the values. I really don't see what the issue is here. It only needs to exist for the postgres case because of a foreign key constraint from the new celery_user_rate_limit table.

jdavcs replied:
Oh, i see - the set difference will exclude any existing ids even if you start with 2. You're right, there's no issue here.

jdavcs commented on lib/galaxy/celery/base_task.py:
```diff
@@ -0,0 +1,163 @@
+        tasks_per_user_per_sec: float,
+        ga_scoped_session: galaxy_scoped_session,
+    ):
+        self.task_exec_countdown_secs = 1 / tasks_per_user_per_sec
```
In the current implementation we only get here if the `celery_user_rate_limit` option is set to a non-zero value. However, this could change. Can we guard against a zero value here? (an assert or handing the exception, etc. - depending on how you decide it would make sense to handle such a case)

claudiofr replied:
I will handle the exception which I believe adds less overhead than doing an assert.

jdavcs commented on test/integration/test_celery_user_rate_limit.py:
```diff
@@ -0,0 +1,169 @@
+        else:
+            secs_between_tasks_per_user = 1 / tasks_per_user_per_sec
+            expected_duration = secs_between_tasks_per_user * (num_calls - 1)
+            expected_duration_lbound = expected_duration - 4
```
I am a little confused about the calculations here. Can you explain, please? (it might be obvious - sorry if that's the case!)
- What is `4` and why is it hardcoded here? (I see it adjusts the lower/higher bounds)
- `secs_between_tasks_per_user = 1 / tasks_per_user_per_sec`: why is that seconds between tasks?

claudiofr replied:
To illustrate, assume tasks_per_user_per_sec = 2. This means that even though the user may submit thousands of tasks to execute at the same time, the user is allowed to execute at most 2 task per second or at most 120 tasks per minute. In this case secs_between_tasks_per_user = 1/2 = .5 sec.
In order to achieve this we need to schedule the tasks via retry to execute .5 seconds apart.

In order to verify that the rate limiting is working as expected I verify that the total time it takes to execute all of the tasks submitted at once in the test execute within a certain period of time. So to continue the above example, if the user submits 120 tasks at once, I need to verify that it takes 60 seconds to complete execution of all the tasks, i.e. secs_between_tasks_per_user(.5) * num_calls(120) = 60 seconds.
However, because I cannot be guaranteed that it will be exactly 60 seconds I make it a range, i.e. 60 + or - 4 seconds. Of course, there is nothing magic about choosing 4 as the degree of error. I came up with it via trial and error. I suppose I could assign it to a class variable.

jdavcs replied:
Thank you - this is very helpful! I think I get it now :) 
The value `4` is fine as is.

jdavcs commented on test/integration/test_celery_user_rate_limit.py:
```diff
@@ -0,0 +1,169 @@
+            found_user_ids = conn.scalars(
+                text("select id from galaxy_user where id between 1 and :high"), {"high": num_users}
+            ).all()
+            if len(expected_user_ids) > len(found_user_ids):
```
With the default value of `num_users` (~~which is what the tests are currently using~~ EDIT: tests are using 3), `len(expected_user_ids)` will be 1. Is that correct? So, when `num_users == 2`, and the database has only one user, this code won't create any additional users, and execute tasks for one user only. I would expect that the number of users tested would equal the value of the `num_users` parameter?

EDIT: I think that's indeed a bug: I've just ran into an integrity error under postgres: it couldn't find user.id=2. Changing to `num_users + 2` in `setup_users` fixed it. However, I haven't verified this - it could be an error in my code too.

claudiofr replied:
It is a bug but it never caused a problem because at the time this code runs there apparently are no users in the galaxy_user table so len(found_user_ids) is always 0 and the condition len(expected_user_ids) > len(found_user_ids) is always true. Apparently the integration testing framework creates a brand new empty database everytime. The test assumes that there will always be a user id = 1, but apparently it is not present at the time this function runs and it somehow gets populated at some point after this code runs. I will change the code to say:
select id from galaxy_user where id between 2 and :high

jdavcs replied:
sometimes a new db is created, sometimes it is not - it depends on how the tests are run. I think, `run_tests.sh` will try to reuse the postgres database, whereas a direct execution via pytest will not, but I may be mistaken. 
The assumption about user.id=1 holds when the test uses the database that has been created by the testing framework. However, when a separate db is created for a test (like when we need to execute under sqlite regardless of the db env var passed to the test in the CI, the default user is not created. So, my suggestion would be to not make any assumptions about existing records in the db.

jdavcs commented on test/integration/test_celery_user_rate_limit.py:
```diff
@@ -0,0 +1,169 @@
+
+
+@galaxy_task(bind=True)
+def mock_user_id_task(self, task_user_id: int):
```
```suggestion
@galaxy_task
def mock_user_id_task(task_user_id: int):
```
No need to bind: we are not using the task instance in the body.

claudiofr replied:
I think this is a very nit picky observation.

jdavcs replied:
It makes the code easier to understand. It's a large code base and debugging (or reading through the code to understand how things work) is often nontrivial. Having only the parameters that are actually used reduced the cognitive load - i.e., less things to keep track of/to consider. Yes, it is a minor comment, but, believe me, we try to hold ourselves to the same standards.

jdavcs commented on test/integration/test_celery_user_rate_limit.py:
```diff
@@ -0,0 +1,169 @@
+        results: Dict[int, List[AsyncResult]] = {}
+        for user in users:
+            user_results: List[AsyncResult] = []
+            for _i in range(num_calls):  # type: ignore
```
```suggestion
            for _ in range(num_calls):  # type: ignore
```
Just use `_` for a throwaway counter variable (pythonic idiom)

claudiofr replied:
fine

jdavcs replied:
same as the comment above: we try to keep the degree of "surprise" to a minimum. It's immediately clear what's happening when the iter variable is an underscore. It's not because it's pythonic - it's for convenience.

jdavcs commented on test/integration/test_celery_user_rate_limit.py:
```diff
@@ -0,0 +1,169 @@
+    def setUp(self):
+        super().setUp()
+
+    def _test_mock_pass_user_id_task(self, users: Iterable[int], num_calls: int, tasks_per_user_per_sec: float):
```
This is a rather complicated function that, I think, could benefit from a few comments (it took me some time to understand what it was doing). Maybe add just a few very brief inline comments to each code block, indicating something like `execute tasks for each user, store results / verify results / verify elapsed time / verify elapsed time per user`? I think that would make it much easier to understand.

claudiofr replied:
In general, I don't like to clutter the code with inline comments. That is why I created a fairly detailed class level comment. Also, I noticed that the galaxy code base is very sparingly commented in general which has made understanding it very slow going. However, I will add the extra comments because I believe in making code as understandable as possible for others.

jdavcs replied:
Thank you! I do believe it'll help. 
(and yes, lots of places in the code base where we should've added more comments, used better variable names, had broken up monster methods, etc.)

jdavcs commented on test/integration/test_celery_user_rate_limit.py:
```diff
@@ -0,0 +1,169 @@
+    @classmethod
+    def handle_galaxy_config_kwds(cls, config):
+        TestCeleryUserRateLimitIntegrationPostgres.handle_galaxy_config_kwds(config)
+        config["celery_user_rate_limit"] = 0.1
```
Since each test class sets the rate to a value and then passes that value to the test function, could you use a class variable instead? Easier to read + less error-prone + immediately clear what's different about that test.

claudiofr replied:
ok

jdavcs commented on lib/galaxy/celery/__init__.py:
```diff
@@ -67,6 +69,21 @@ def trim_module_name(self, module):
+        """
+        app = get_galaxy_app()
+        assert app
+        app[GalaxyTaskBeforeStart](self, task_id, args, kwargs)
```
I think it would be more intuitive if you called a method here instead of the object itself. Consider this:
```py
pretask = app[GalaxyTaskBeforeStart]
pretask.check_rate_limit(self, kwargs)   # Also, you don't need task_id and kwargs for that method
```
I think an explicit call to a method with a descriptive name makes it immediately clear what's going on. As a side benefit, it offers flexibility: we can call other methods if we need to. (For the base object, it would be a noop, just like the `__call__` implementation)

claudiofr replied:
This was a deliberate design decision. The GalaxyTaskBeforeStart was intended as the class that would have all of the before start logic. If we needed to add more functionality it would be done within the GalaxyTaskBeforeStart.__call__ method without having to change the actual GalaxyTask.before_start method. So we either place the calls to the lower level distinct operations in the GalaxyTask.before_start or in GalaxyTaskBeforeStart.__call__ method. Because we might add other logic in the future I decided to pass all available information including the task_id and args. I see this as design preference.

jdavcs commented on lib/galaxy/app.py:
```diff
@@ -581,8 +586,30 @@ def __init__(self, configure_logging=True, use_converters=True, use_display_appl
+        """
+        if self.config.celery_user_rate_limit:
+            task_before_start: GalaxyTaskBeforeStart
+            if (
```
We have a utility method `is_postgres` that does exactly that, but is located in the [wrong place](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/model/unittest_utils/model_testing_utils.py#L224). My suggestion is to move that method into `galaxy.model.database_utils` (there's only [one other place](https://github.com/galaxyproject/galaxy/blob/dev/test/unit/app/test_migrate_database.py#L11) where it's used - it would need to be adjusted), and use it here. Given that you wouldn't have the `celery_user_rate_limit_standard_before_start` config option, I think this code would become even simpler:
```py
if is_postgres(self.config.database_connection):
    task_before_start = GalaxyTaskBeforeStartUserRateLimitPostgres(
        self.config.celery_user_rate_limit, self.model.session
    )
else:
    task_before_start = GalaxyTaskBeforeStartUserRateLimitStandard(
        self.config.celery_user_rate_limit, self.model.session
    )
```

claudiofr replied:
I will make changes to use is_postgres. I would like to keep the celery_user_rate_limit_standard_before_start option. However, I guess I could never really see a need for it if we are using postgres as a database.

jdavcs commented on lib/galaxy/celery/base_task.py:
```diff
@@ -0,0 +1,163 @@
+            count_down = next_scheduled_time - now
+            task.retry(countdown=count_down.total_seconds())
+
+    @abstractmethod
```
I don't think you need this decorator since it's not a really an abstract method since it has a concrete implementation (unless you intended it as one, in which case it doesn't need to return `now`, I guess?)

claudiofr replied:
It is meant to be an abstract method. It returns now because one of the linters in the CI complained about it when it did not return anything. Other than the linter complaining the code worked fine when the abstract method did not return anything.

jdavcs commented on test/integration/test_celery_user_rate_limit.py:
```diff
@@ -0,0 +1,169 @@
+class TestCeleryUserRateLimitIntegrationPostgres1(TestCeleryUserRateLimitIntegrationPostgres):
+    @classmethod
+    def handle_galaxy_config_kwds(cls, config):
+        TestCeleryUserRateLimitIntegrationPostgres.handle_galaxy_config_kwds(config)
```
For all calls to the parent class, better use `super().your-call-here()` vs. specifying the parent.

claudiofr replied:
I'll make the change. For some reason I thought I tried that initially and it didn't work as expected.

claudiofr committed:
Pass user id to celery tasks to enable new user rate limiting functionality.

Remove unnecessary integration tests. Change calls to celery tasks to pass values
to new task_user_id parameter which is used for user rate limiting.

claudiofr committed:
Pass user id to celery tasks to enable new user rate limiting functionality.

Fix signature to overriden PurgableManagerMixin.purge method. Rather
than having an explicit user parameter assume it is in the **kwargs param.

claudiofr committed:
Pass user id to celery tasks to enable new user rate limiting functionality.

Added # type: ignore[arg-type] for dburl parameter to is_posgress function.

jdavcs committed:
Remove unused config property

jdavcs reviewed (approved):
Thank you, @claudiofr, for your contribution and for addressing all the comments! This is great work!
